R/LoadReporterModule.R

```diff
@@ -0,0 +1,140 @@
+#' Load Report User Interface
+#' @description `r lifecycle::badge("experimental")`
+#' button for adding views/cards to the Report.
+#'
+#' For more details see the vignette: `vignette("simpleReporter", "teal.reporter")`.
+#' @param id `character(1)` this `shiny` module's id.
+#' @return `shiny::tagList`
+#' @export
+report_load_ui <- function(id) {
+  ns <- shiny::NS(id)
+
+  shiny::tagList(
+    shiny::singleton(
+      shiny::tags$head(shiny::includeCSS(system.file("css/custom.css", package = "teal.reporter")))
+    ),
+    shiny::tags$button(
+      id = ns("reporter_load"),
+      type = "button",
+      class = "simple_report_button btn btn-primary action-button",
+      title = "Load",
+      `data-val` = shiny::restoreInput(id = ns("reporter_load"), default = NULL),
+      NULL,
+      shiny::tags$span(
+        shiny::icon("upload")
+      )
+    )
+  )
+}
+
+#' Load Report Server
+#' @description `r lifecycle::badge("experimental")`
+#' server for saving and loading to the Report.
+#'
+#' For more details see the vignette: `vignette("simpleReporter", "teal.reporter")`.
+#'
+#' @param id `character(1)` this `shiny` module's id.
+#' @param reporter [`Reporter`] instance.
+#'
+#' @return `shiny::moduleServer`
+#' @export
+report_load_srv <- function(id, reporter) {
+  checkmate::assert_class(reporter, "Reporter")
+
+  shiny::moduleServer(
+    id,
+    function(input, output, session) {
+      ns <- session$ns
+
+      archiver_modal <- function() {
+        nr_cards <- length(reporter$get_cards())
+        shiny::modalDialog(
+          easyClose = TRUE,
+          shiny::tags$h3("Load the Report"),
+          shiny::tags$hr(),
+          shiny::fileInput(ns("archiver_zip"), "Choose Archiver File to Load (a zip file)",
+            multiple = FALSE,
+            accept = c(".zip")
+          ),
+          footer = shiny::div(
+            shiny::tags$button(
+              type = "button",
+              class = "btn btn-danger",
+              `data-dismiss` = "modal",
+              `data-bs-dismiss` = "modal",
+              NULL,
+              "Cancel"
+            ),
+            shiny::tags$button(
+              id = ns("reporter_load_main"),
+              type = "button",
+              class = "btn btn-primary action-button",
+              `data-val` = shiny::restoreInput(id = ns("reporter_load_main"), default = NULL),
```

Comment:
This is controversial. Probably you don't want to restoreInput of a clicked button. This will probably trigger observer and possibly cause an error. Lets analyse what would happen when restoring:
- button will trigger observer where `load_json_report(reporter, input$archiver_zip[["datapath"]], input$archiver_zip[["name"]])` is executed (probably both inputs will be restored)
- then or before previewer will try to add cards from the `state$values`

Reply:
it is a default shiny setup for actionButton

```
> shiny::actionButton
function (inputId, label, icon = NULL, width = NULL, ...) 
{
    value <- restoreInput(id = inputId, default = NULL)
    tags$button(id = inputId, style = css(width = validateCssUnit(width)), 
        type = "button", class = "btn btn-default action-button", 
        `data-val` = value, list(validateIcon(icon), label), 
        ...)
}
<bytecode: 0x12e965348>
<environment: namespace:shiny>
```

Reply:
I use tags$button to have better control.

Reply:
Please add `setBookmarkExclude` for all action buttons in this PR and don't use `restoreInput` for them

Reply:
Reproducible stages:

1. Run the app, add and download report
2. Load a report
3. Click bookmark and open a link Load-dialog will be opened

<details>
<summary>Example app</summary>

```r
library(shiny)
library(teal.reporter)
library(ggplot2)
library(rtables)
library(DT)
library(bslib)


# any of c("add", "download", "load", "reset")
options("teal.reporter.simple_reporter_modules" = c("add", "download", "load", "reset"))

ui <- fluidPage(
    # please, specify specific bootstrap version and theme
    theme = bs_theme(version = "4"),
    titlePanel(""),
    tabsetPanel(
        tabPanel(
            "main App",
            tags$br(),
            sidebarLayout(
                sidebarPanel(
                    uiOutput("encoding")
                ),
                mainPanel(
                  bookmarkButton(),
                    tabsetPanel(
                        id = "tabs",
                        tabPanel("Plot", plotOutput("dist_plot")),
                        tabPanel("Table", verbatimTextOutput("table")),
                        tabPanel("Table DataFrame", verbatimTextOutput("table2")),
                        tabPanel("Table DataTable", dataTableOutput("table3"))
                    )
                )
            )
        ),
        ### REPORTER
        tabPanel(
            "Previewer",
            reporter_previewer_ui("prev")
        )
        ###
    )
)
server <- function(input, output, session) {
    output$encoding <- renderUI({
        tagList(
            ### REPORTER
            teal.reporter::simple_reporter_ui("simple_reporter"),
            ###
            if (input$tabs == "Plot") {
                sliderInput(
                    "binwidth",
                    "binwidth",
                    min = 2,
                    max = 10,
                    value = 8
                )
            } else if (input$tabs %in% c("Table", "Table DataFrame", "Table DataTable")) {
                selectInput(
                    "stat",
                    label = "Statistic",
                    choices = c("mean", "median", "sd"),
                    "mean"
                )
            } else {
                NULL
            }
        )
    })
    plot <- reactive({
        req(input$binwidth)
        x <- mtcars$mpg
        ggplot(data = mtcars, aes(x = mpg)) +
            geom_histogram(binwidth = input$binwidth)
    })
    output$dist_plot <- renderPlot(plot())
    
    table <- reactive({
        req(input$stat)
        lyt <- basic_table() %>%
            split_rows_by("Month", label_pos = "visible") %>%
            analyze("Ozone", afun = eval(str2expression(input$stat)))
        build_table(lyt, airquality)
    })
    output$table <- renderPrint(table())
    
    table2 <- reactive({
        req(input$stat)
        data <- aggregate(
            airquality[, c("Ozone"), drop = FALSE], list(Month = airquality$Month), get(input$stat),
            na.rm = TRUE
        )
        colnames(data) <- c("Month", input$stat)
        data
    })
    output$table2 <- renderPrint(print.data.frame(table2()))
    output$table3 <- renderDataTable(table2())
    
    ### REPORTER
    reporter <- Reporter$new()
    card_fun <- function(card = ReportCard$new(), comment) {
        if (input$tabs == "Plot") {
            card$set_name("Plot Module")
            card$append_text("My plot", "header2")
            card$append_plot(plot())
            card$append_rcode(
                paste(
                    c(
                        "x <- mtcars$mpg",
                        "ggplot2::ggplot(data = mtcars, ggplot2::aes(x = mpg)) +",
                        paste0("ggplot2::geom_histogram(binwidth = ", input$binwidth, ")")
                    ),
                    collapse = "\n"
                ),
                echo = TRUE,
                eval = FALSE
            )
        } else if (input$tabs == "Table") {
            card$set_name("Table Module rtables")
            card$append_text("My rtables", "header2")
            card$append_table(table())
            card$append_rcode(
                paste(
                    c(
                        "lyt <- rtables::basic_table() %>%",
                        'rtables::split_rows_by("Month", label_pos = "visible") %>%',
                        paste0('rtables::analyze("Ozone", afun = ', input$stat, ")"),
                        "rtables::build_table(lyt, airquality)"
                    ),
                    collapse = "\n"
                ),
                echo = TRUE,
                eval = FALSE
            )
        } else if (input$tabs %in% c("Table DataFrame", "Table DataTable")) {
            card$set_name("Table Module DF")
            card$append_text("My Table DF", "header2")
            card$append_table(table2())
            # Here r code added as a regular verbatim text
            card$append_text(
                paste0(
                    c(
                        'data <- aggregate(airquality[, c("Ozone"), drop = FALSE], list(Month = airquality$Month), ',
                        input$stat,
                        ", na.rm = TRUE)\n",
                        'colnames(data) <- c("Month", ', paste0('"', input$stat, '"'), ")\n",
                        "data"
                    ),
                    collapse = ""
                ), "verbatim"
            )
        }
        if (!comment == "") {
            card$append_text("Comment", "header3")
            card$append_text(comment)
        }
        card
    }
    teal.reporter::simple_reporter_srv("simple_reporter", reporter = reporter, card_fun = card_fun)
    teal.reporter::reporter_previewer_srv("prev", reporter)
}

shinyApp(ui = ui, server = server, enableBookmarking = "server")


```
</details>

Reply:
setBookmarkExclude added to LoadModule and appended in Previewer:)